### DISCUSSION TITLE
Looking at Lua Documentation formats

### DISCUSSION THREAD

**Discussion by lewri** [CorsixTH/CorsixTH#2751] [Research Room]
I know a while ago there has been some discussion by @Alberth289346 with othwes in our chat channels about our Lua documentation system ( https://github.com/CorsixTH/CorsixTH/wiki/CodeDocumentationStyle ) . I did find out the other day it looks like our documentation system may be bespokely written by Corsix ( https://github.com/CorsixTH/CorsixTH/blob/master/LDocGen/main.lua ) in order to generate the CorsixTH Lua Documentation guide: http://corsixth.com/CorsixTH/corsixth_lua/index.html

This generated documentation still currently remains a little broken (#1793) in the main repo, though forked versions seem to function slightly better.

I wanted to open the discussion more so that the documentation methods can be looked at, and decided upon if any changes are necessary. I believe I remember at one point a nice goal was integrating it into Visual Studio Code with some supported extension.

There is also LDoc, which looks like it has some simmers of development still but not highly active: https://github.com/lunarmodules/ldoc

**Comment by Alberth289346** [comment]
The Lua editor plugins in VScode come with an assumed documentation format. In other words, there is no choice after picking a plugin.

If I remember correctly, there are 2 main contestants for a Lua plugin (not sure how much alive they are):
- official lua (not sure what Lua version exactly).
- luau (a 5.1 compatible independent derivative, with some but not all of the features of newer luas and some features that do not exist not in official lua at all).

Likely it needs an evaluation of the relevant plugins to understand what we want to have.

----
While experimenting with VScode, I found that editing CTH code with our documentation in such a plugin is a non-trivial problem. The reason is our classes which do not fit with assumptions of the plugins.  There is also lack of much type information in our code. I under-estimated the impact of that.
I mostly tried to find a fast way to convince the editor to accept our code mostly as-is. That didn't work.

A more careful small-scale attempt is probably more useful here. Eg `class.lua`, `strict.lua` and a file with a class. Then try to make that work. Second, the C++ functionality would need to be recognized. I didn't look into that at all.

----
Separate from allowing to use VScode to be used, we could switch to a different/modern luadoc system. Our ldoc parser is quite broken in the sense it just crashes if your lua or lua doc is not correct, without useful message. Improving that would already be valuable.

I am not sure what requirements sch a documentation system would need to have.

Fwiw, I wrote a Lua 5.4 parser in Java some time ago, and that also understands our lua doc. (Not sure I completely finished it but that can be done if desired. Other extensions are of course also possible.)
It beats the current ldoc implementation in not crashing and in also finding completely undocumented functions.  
On the other hand, it doesn't produce any nicely formatted output at all, and it's code that nobody else uses or even knows about. From a CTH point of view it's more work to maintain, compared with picking some other lua doc system that is still active.

**Comment by TheCycoONE** [comment]
I use the Lua Language Server plugin for NeoVim and Code (it's also available for JetBrains products): https://luals.github.io/ - it's a healthy, actively maintained project, with a lot of use.

It seems to use it's own annotation system they call LuaCATS which is an extension of EmmyLua. It has it's own documentation generator but I understand it is in poor shape at best. There is also a community project to generate documentation with those annotations but it doesn't have much support behind it: https://github.com/Ottatop/lcat

The standard mainstays of lua documentation generators ldoc and luadoc do not use a compatible syntax. We may be better off adopting the mainstream language server and it's annotations and extending our doc generator to compensate?
